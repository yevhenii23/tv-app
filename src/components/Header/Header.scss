$mainColor: #3F3F3F;
$buttonColor: #48958B;
$textColor: #F9FAFF;

.header {
    background-color: $mainColor;
    display: flex;
    justify-content: space-between;
    align-items: center;
    height: 8vh;
    .text {
        color: $textColor;
    }
    .heading {
        margin-left: 10px;
    }
}
.label {
    margin-right: 10px;
    .link {
        color:$buttonColor;
        margin-right: 10px;
        text-decoration: none;
    }
    span {
        margin-right: 15px;
        color:red;
    }
}

input[type=button] {
    background-color: $buttonColor;
    border:none;
    color: $textColor;
    text-decoration: none;
    border-top-right-radius: 10px;
    border-bottom-right-radius: 10px;
    height: 35px;
    width: 5vw;
}

input[type=search] {
    border-bottom-left-radius: 10px;
    border-top-left-radius: 10px;
    height: 35px;
    width: 20vw;
}