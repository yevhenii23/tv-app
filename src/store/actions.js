import axios from 'axios';

export const FETCH_MOVIE_LIST = 'FETCH_MOVIE_LIST';
export const FETCH_MOVIE_LIST_SUCCESS = 'FETCH_MOVIE_LIST_SUCCESS';
export const ADD_TO_MY_LIST = 'ADD_TO_MY_LIST';
export const DELETE_FROM_MY_LIST = 'DELETE_FROM_MY_LIST';
export const CHANGE_SEARCH_TEXT ='CHANGE_SEARCH_TEXT';

const getUrl = (searchText) => searchText ? 'http://api.tvmaze.com/search/shows?q=' + searchText : 'http://api.tvmaze.com/shows';

export const fetchMovies = () => (dispatch, getState) => {
    const { searchText } = getState();
    const url = getUrl(searchText);

    dispatch({ type:FETCH_MOVIE_LIST });
    axios.get(url)
        .then(({ data: list }) => {
            dispatch({
                type:FETCH_MOVIE_LIST_SUCCESS,
                list: searchText ? list.map(res => res.show) : list,
            })
        })
}

export const addToFavorites = (id) => {
    return {
        type: ADD_TO_MY_LIST,
        id
    }
}

export const changeSearchText = (text) => {
    return {
        type: CHANGE_SEARCH_TEXT,
        text
    }
}

export const deleteFromFavorites = (id) => {
    return {
        type: DELETE_FROM_MY_LIST,
        id
    }    
}