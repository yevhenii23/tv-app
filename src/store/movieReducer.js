import { 
    FETCH_MOVIE_LIST, ADD_TO_MY_LIST,
    FETCH_MOVIE_LIST_SUCCESS, CHANGE_SEARCH_TEXT,
    DELETE_FROM_MY_LIST } from './actions'

export const initialState = {
    list: [],
    searchText: '',
    favorites: [],
    loading: false,
};

const getUniq = array => Array.from(new Set(array));

export function movieReducer (state = initialState, action) {
    switch(action.type) {
        case FETCH_MOVIE_LIST:
            return {
                ...state,
                loading: true,                
            };
        case FETCH_MOVIE_LIST_SUCCESS:
            return {
                ...state,
                loading: false,
                list: action.list,
            };
        case CHANGE_SEARCH_TEXT:
            return {
                ...state,                
                searchText: action.text,
            };
        case ADD_TO_MY_LIST:
            return {
                ...state,
                favorites: getUniq([
                    ...state.favorites,
                    action.id
                ])
            };
        case DELETE_FROM_MY_LIST: 
            return {
                ...state,
                favorites: state.favorites.filter(id => id !== action.id)
            };
            
       default: 
           return state;
    }
}
