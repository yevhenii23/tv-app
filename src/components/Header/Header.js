import React from 'react';
import './Header.scss';
import { Link } from "react-router-dom";
import { connect } from 'react-redux';
import { fetchMovies, changeSearchText } from '../../store/actions';

const Header = ({ favoritesLength, searchText, getMovies, changeSearch }) => (

    <header data-test="header-test" className="header"> 
        <h1 data-test="h1-logo-test" className="text heading">TV-APP</h1>
        <div data-test="wraper-label-test" className="label">
            <Link data-test="link-to-home-test" to="/" className="link">Home</Link>
            <Link data-test="link-to-myList" to="/mylist" className="link">My list:</Link>
            <span data-test="counter-test">{favoritesLength}</span>  
            <input data-test="search-test" type="search" placeholder="Type to Search"  value={searchText} onChange={changeSearch} className="search-input"></input>
            <input data-test="search-button-test" type="button" value="Search" onClick={getMovies} className="button-input"></input>
        </div>
    </header>
);


const mapStateToProps = ({ favorites, searchText }) => {
    return {
        favoritesLength: favorites.length,
        searchText: searchText,
    }
}

const mapDispatchToProps = dispatch => ({
    getMovies: () => dispatch(fetchMovies()),
    changeSearch: (e) => dispatch(changeSearchText(e.target.value)),
});

export default connect(mapStateToProps, mapDispatchToProps)(Header);