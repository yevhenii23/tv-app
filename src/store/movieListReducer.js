import { FETCH_MOVIE_LIST, ADD_TO_MY_LIST, ADD_ABOUT_INFO } from './actions'

export const initialState = {
    fetchedList: [],
    favorites: [],
    about:'',
};


export function movieListReducer (state = initialState, action) {
    switch(action.type) {
        case FETCH_MOVIE_LIST: 
            return {...state,fetchedList:action.payload}
        case ADD_TO_MY_LIST:
            return {...state,favorites:[...state.favorites,action.payload]}
        case ADD_ABOUT_INFO:
            return {...state, about:action.payload }
       default: 
           return state;
    }
}
