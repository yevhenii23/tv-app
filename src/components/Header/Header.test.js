import React from 'react';
import { shallow } from 'enzyme';
import Header from './Header';
import { fetchMovies } from '../../store/actions';


// замена того что прилетит с импорта на подельный код мока
jest.mock('react-redux', () => {
    const dispatch = jest.fn();
    const state = { 
        favorites: [1,2,3,4,5],
        searchText: 'aaa' 
    };
    const mockConnect = (mapState, mapDispatch) => (Component) => (ownProps) => {
        const props = {
            ...ownProps,
            ...(mapState(state)),
            ...(mapDispatch(dispatch))
        };
        return <Component {...props}/>   
    }
    return {
        connect: mockConnect,
    }; 
});
jest.mock('../../store/actions');

describe('Header component', () => {
    let wrapper;

    beforeEach(() => {
        wrapper = shallow(<Header />).dive();
    });

    test('render without errors', () => {
        expect(wrapper).toMatchSnapshot();
    })

    test('calls getMovies on search button click ', () => {
        wrapper.find('input.button-input').simulate('click');
        expect(fetchMovies).toHaveBeenCalled();
    })
});