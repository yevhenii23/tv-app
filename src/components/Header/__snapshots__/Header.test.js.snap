// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Header component render without errors 1`] = `ShallowWrapper {}`;
